src/sys/windows.rs

```diff
@@ -398,24 +398,30 @@ pub fn open_stdin_if_redirected() -> Option<File> {
 }
 
 /// A unique identifier for a file.
-#[derive(Clone)]
-#[repr(transparent)]
-pub struct FileId(FileSystem::FILE_ID_INFO);
+pub enum FileId {
+    Id(FileSystem::FILE_ID_INFO),
+    Path(PathBuf),
+}
 
 impl PartialEq for FileId {
     fn eq(&self, other: &Self) -> bool {
-        // Lowers to an efficient word-wise comparison.
-        const SIZE: usize = std::mem::size_of::<FileSystem::FILE_ID_INFO>();
-        let a: &[u8; SIZE] = unsafe { mem::transmute(&self.0) };
-        let b: &[u8; SIZE] = unsafe { mem::transmute(&other.0) };
-        a == b
+        match (self, other) {
+            (FileId::Id(left), FileId::Id(right)) => {
+                // Lowers to an efficient word-wise comparison.
+                const SIZE: usize = std::mem::size_of::<FileSystem::FILE_ID_INFO>();
+                let a: &[u8; SIZE] = unsafe { mem::transmute(left) };
+                let b: &[u8; SIZE] = unsafe { mem::transmute(right) };
+                a == b
+            }
+            (FileId::Path(left), FileId::Path(right)) => left == right,
+            _ => false,
```

Comment:
this assumes that all files which can be referenced by ID are. we will never try to resolve a path against an ID.